phoenix-core/src/main/java/org/apache/phoenix/util/IndexUtil.java

```diff
@@ -849,4 +860,80 @@ public static long getIndexPendingDisableCount(PhoenixConnection conn, String fa
             throw new IOException(e);
         }
     }
+
+    private static boolean containsOneOrMoreColumn(Scan scan) {
+        Map<byte[], NavigableSet<byte[]>> familyMap = scan.getFamilyMap();
+        if (familyMap == null || familyMap.isEmpty()) {
+            return false;
+        }
+        for (Map.Entry<byte[], NavigableSet<byte[]>> entry : familyMap.entrySet()) {
+            NavigableSet<byte[]> family = entry.getValue();
+            if (family != null && !family.isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static void addEmptyColumnToScan(Scan scan, byte[] emptyCF, byte[] emptyCQ) {
+        boolean addedEmptyColumn = false;
+        Iterator<Filter> iterator = ScanUtil.getFilterIterator(scan);
+        while (iterator.hasNext()) {
+            Filter filter = iterator.next();
+            if (filter instanceof EncodedQualifiersColumnProjectionFilter) {
+                ((EncodedQualifiersColumnProjectionFilter) filter).addTrackedColumn(ENCODED_EMPTY_COLUMN_NAME);
+                if (!addedEmptyColumn && containsOneOrMoreColumn(scan)) {
+                    scan.addColumn(emptyCF, emptyCQ);
+                }
+            }
+            else if (filter instanceof ColumnProjectionFilter) {
+                ((ColumnProjectionFilter) filter).addTrackedColumn(new ImmutableBytesPtr(emptyCF), new ImmutableBytesPtr(emptyCQ));
+                if (!addedEmptyColumn && containsOneOrMoreColumn(scan)) {
+                    scan.addColumn(emptyCF, emptyCQ);
+                }
+            }
+            else if (filter instanceof MultiEncodedCQKeyValueComparisonFilter) {
+                ((MultiEncodedCQKeyValueComparisonFilter) filter).setMinQualifier(ENCODED_EMPTY_COLUMN_NAME);
+            }
+            else if (!addedEmptyColumn && filter instanceof FirstKeyOnlyFilter) {
+                scan.addColumn(emptyCF, emptyCQ);
+                addedEmptyColumn = true;
+            }
+        }
+        if (!addedEmptyColumn && containsOneOrMoreColumn(scan)) {
+            scan.addColumn(emptyCF, emptyCQ);
+        }
+    }
+
+    public static void setScanAttributes(Scan scan, PTable table, PhoenixConnection phoenixConnection) throws SQLException {
```

Comment:
Nit: Can we rename this method to setScanAtributesForIndexRepair or similar? Also, wondering why the method is in IndexUtil vs Scan or IndexMaintainer.

Reply:
I will rename the method. It can be part of IndexUtil as it is only for index tables and sets index metadata, or can be part of ScanUtil as it changes scan attributes. I chose IndexUtil.